### ISSUE TITLE
New Toms

### ISSUE THREAD

**Issue by sneakernets** [sneakernets/DMXOPL#34]
I may have to replace the toms with the older ones, as the new ones I just can't figure out how to make louder....

But I would like to salvage the idea, somehow, and make it work -- _somehow_.

**Comment by freq-mod** [comment]
New toms are great, if only to resolve volume issue...
Try rising Carrier 1 operator volume to 63 (at least on floor toms)
OR you may try something similar to this:
![tomz](https://user-images.githubusercontent.com/32672779/34166067-45334c02-e4de-11e7-8eac-172f166948eb.PNG)

Recently I was making a GM patch set for use with libOPNMIDI (Sega Genesis audio chip MIDI play system). I had problem with quiet drums, I """"resolved""""" it by lowering volume of all melodic instruments. So it might be a last resort, being as idiotic as it sounds.

**Comment by sneakernets** [comment]
Yeah I was thinking of that too, but the real problem is lack of....well, let me put it this way, OPL3 doesn't have SSG-EG features. I'm not really sure how one could make a convincing toms for OPL3 without using pitch slides, and there's no way to do that in DMX (or WOPL, for that matter).

It gives me a sad!

**Comment by freq-mod** [comment]
I didn't use SSG-EG in my OPN2 bank (frankly speaking, I hardly ever use that). Problems are:
1) DMX doesn't support 4-op, only dual 2-op
2) OPL3's 4-op mode is shit - it lacks half of OPN2 algorithms, one of them (algorithm 3) is very useful for guitars, basses, flutes and... tom-tom drums
3) There is most likely no way to implement MIDI-independent pitch slides (but one would have to ask Wohlstand, he knows more about it lol)

**Comment by Wohlstand** [comment]
> There is most likely no way to implement MIDI-independent pitch slides (but one would have to ask Wohlstand, he knows more about it lol)

You can see that vibrato controller is works on libADLMIDI side (it's real vibrato which is same as on most WT synths, I'm not mean per-operator vibrato flag and I don't mean same way as Jammie O'Connel's way to turn on tremolo flag on instrument on high vibrato CC level) and it's a MIDI-independent pitch slide. Currently there are three pitch slides:
* MIDI-pitch
* Vibrato pitch
* Second 2-op voice detune

Then will be a four slide which will implement similar to SF2 feature where are attack / sustain and decay based on the pitch rather volume. Seems a good idea, I would to try to implement this. It will be on both OPL3 and OPN2 libraries as I using same MIDI playing code on both libraries.

P.S. Let's dump output of those thing together with XG's CC74 into IMF and give to someone who know how to create IMF files (by old ways) to let him to wonder :-P

**Comment by sneakernets** [comment]
> OPL3's 4-op mode is shit

Ain't telling me anything I don't already know! I discovered that most 4-op instruments on OPL-3 I could recreate with two 2-op voices, which is "technically 4-op" if you wanna be cheeky.

> Then will be a four slide which will implement similar to SF2 feature where are attack / sustain and decay based on the pitch rather volume.

😮

**Comment by freq-mod** [comment]
So technically, it would be possible to have sth like that?
https://instaud.io/1yyb

**Comment by Wohlstand** [comment]
> Ain't telling me anything I don't already know! I discovered that most 4-op instruments on OPL-3 I could recreate with two 2-op voices, which is "technically 4-op" if you wanna be cheeky.

Some algorithms still be impossible even with pseudo-4op, for example the `1 => (2 + 3 + 4)`. I have tried to port some OPN2 instruments to OPL3 into my bank and I have found lack of algorithm, therefore I have tweaked to make the sound be similar with a different algorithm. As example, the harpsichord and acoustic bass in my 68'th bank. OPL3 also lacks second decay which is on OPN2. However, OPL3 gives 8 different waveforms per operator unlike OPN2 with sine-only operators. Also, OPL3's key scale works for velocity too while on OPN2 it affects attack and decay timing only.

> So technically, it would be possible to have sth like that?

Yeah, and this will go like a software-side trick which will work even on the real chip. As this will be a library-side trick, nothing need to write in the MIDI file to make this.

**Comment by freq-mod** [comment]
Is it still an issue?
Because I **think** I might have accidentally fixed that for good

**Comment by sneakernets** [comment]
I just heard those toms and yeah they pretty good!

I'm still doing some testing though, been a little quiet lately since January is hectic at work!

**Comment by sneakernets** [comment]
Okay, Now that DMX side of things is "done", I can make some 4-op instruments for sure. That should fix the problem.

I know I said this last time!!!
